File: leetcode/340-longest-substring-with-at-most-k-distinct-characters.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var lengthOfLongestSubstringKDistinct = function(s, k) {
  if (k <= 0) return 0;

  var n = s.length;
  if (k >= n) return n;

  var maxlen = 0;
  var len = {}, c;
  for (var i = 0, j = 0; j < n; ++j) {
    c = s[j];
    if (!(c in len)) len[c] = 0;
    ++len[c];
    while (Object.keys(len).length > k) {
      c = s[i++];
      if (0 === --len[c])
        delete len[c];
    }
    if (j - i + 1 > maxlen)
      maxlen = j - i + 1;
  }
  return maxlen;
};

File: leetcode/340-longest-substring-with-at-most-k-distinct-characters.cpp
class Solution {
 public:
  int lengthOfLongestSubstringKDistinct(string s, int k)
  {
    if (k <= 0) return 0;
    if (k >= s.size()) return s.size();

    int n = s.size();
    unordered_map<char, int> cnt;
    int maxlen = 0;
    for (int i = 0, j = 0; j < n; ++j) {
      cnt[s[j]]++;
      while (cnt.size() > k) {
        char c = s[i++];
        if (0 == --cnt[c])
          cnt.erase(c);
      }
      if (j - i + 1 > maxlen)
        maxlen = j - i + 1;
    }
    return maxlen;
  }
};
